src/workspacer.Sound/SoundManager.cs

```diff
@@ -0,0 +1,393 @@
+            });
+
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceRemoved DataFlow:[{dataFlow}] DeviceRole:[{deviceRole}] DeviceId:[{defaultDeviceId}]");
+#endif
+        }
+
+        public void OnDeviceStateChanged(string deviceId, uint newState)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh() );
+
+            DeviceChanged?.Invoke(new DeviceInfo { Id = deviceId }, newState);
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceStateChanged deviceId:[{deviceId}] newState:[{newState}]");
+#endif
+        }
+
+        public void OnDeviceAdded(string deviceId)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh());
+
+            DeviceAdded?.Invoke(new DeviceInfo { Id = deviceId });
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceAdded DeviceId:[{deviceId}]");
+#endif
+        }
+
+        public void OnDeviceRemoved(string deviceId)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh());
+
+            DeviceRemoved?.Invoke(new DeviceInfo { Id = deviceId });
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceRemoved DeviceId:[{deviceId}]");
+#endif
+        }
+
+        public void OnPropertyValueChanged(string deviceId, PROPERTYKEY propertyKey)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh());
+
+            DevicePropertyValueChanged?.Invoke(new DeviceInfo { Id = deviceId }, propertyKey);
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceStateChanged deviceId:[{deviceId}] propertyKey:[{propertyKey}]");
+#endif
+        }
+        #endregion
+    }
+
+    public class SoundClient : IAudioEndpointVolumeCallback
+    {
+        public event DeviceDataEventDelegate<float> VolumeChanged;
+        public event DeviceDataEventDelegate<bool> MuteChanged;
+
+        private IMMDevice _device;
+
+        public DeviceInfo DeviceInfo { get; set; }
+
+        private IAudioEndpointVolume _deviceAudioEndpointVolume;
+
+        public void SetDevice(IMMDevice device)
+        {
+            _device = device;
+            InitializeAudioEnpoint(device);
+        }
+
+        private void InitializeAudioEnpoint(IMMDevice device = null)
+        {
+            device ??= _device;
+
+            // Fetch IAudioEndpointVolume
+            object volObject;
+            device.Activate(new Guid(ComIIDs.IAudioEndpointVolumeIID), (uint)CLSCTX.CLSCTX_INPROC_SERVER, IntPtr.Zero, out volObject);
+            if (volObject is IAudioEndpointVolume)
+            {
+                _deviceAudioEndpointVolume = volObject as IAudioEndpointVolume;
+                _deviceAudioEndpointVolume.RegisterControlChangeNotify(this);
+            }
+        }
+
+        //Notify on default device change
+        public int OnNotify(IntPtr notificationData)
+        {
+
+#if EVENT_DEBUG
+            MessageBox.Show($"OnNotify {notificationData}");
+#endif
```

Comment:
These ifdefs is probably not going to make it into git master, right? :smile:

Reply:
Will remove! To be honest I forgot about these after testing